File: triangle.rb
class Triangle
  attr_reader :t_sides
  def initialize(t_sides)
    @t_sides = t_sides
  end
  def valid_triangle?
    t_sides.any? {|side| side <= 0} || t_sides.min(2).sum < t_sides.max
  end
  def equilateral?
    return false if valid_triangle?
    t_sides.all? {|side| t_sides.count(side) == 3}
  end
  def isosceles?
    return false if valid_triangle?
    t_sides.any? {|side| t_sides.count(side) >= 2}
  end
  def scalene?
    return false if valid_triangle?
    t_sides.all? {|side| t_sides.count(side) == 1}
  end
end

File: javascript/triangle/triangle.js
function Triangle(a, b, c) {
  this.a = a;
  this.b = b;
  this.c = c;
}

Triangle.prototype.kind = function() {
  if (this.a + this.b <= this.c || this.a + this.c <= this.b ||this.b + this.c <= this.a) {
    throw new Error('Triangle violates triangle inequality');
  }

  if (this.a === this.b && this.a === this.c) {
    return 'equilateral';
  }

  if (this.a === this.b || this.a === this.c || this.b === this.c) {
    return 'isosceles';
  } else {
    return 'scalene';
  }
}

export { Triangle };
